torch/cuda/__init__.py

```diff
@@ -60,7 +60,37 @@
         if not _version.hip:
             import pynvml  # type: ignore[import]
         else:
-            import amdsmi  # type: ignore[import]
+            import ctypes
+
+            _original_CDLL = ctypes.CDLL  # type: ignore[misc,assignment]
+
+            # In ROCm there are 2 copies of libamd_smi.so:
+            # - One at lib/libamd_smi.so
+            # - One at share/amd_smi/amdsmi/libamd_smi.so
+            #
+            # The amdsmi python module hardcodes loading the second one in share.
+            # This creates an ODR violation if the copy of libamd_smi.so from lib
+            # has been loaded previously.
+            #
+            # In order to avoid this violation we hook CDLL and try using the
```

Comment:
Would be good to add a permalink to https://github.com/ROCm/amdsmi/blob/9872083b491c9b136496fd493414aca5c6c144cb/py-interface/amdsmi_wrapper.py#L168 to explain why this logic is needed here